### PR TITLE
[inductor] Fix frexp returning mantissa 0 for float32 subnormals on CUDA

### LINKED ISSUE #195007: [torch.compile][Inductor] frexp returns a mantissa of 0 for float32 subnormals on CUDA
## Summary

`torch.frexp` under `torch.compile` gives a mantissa of 0 for float32 subnormals on CUDA. Eager gets it right, so you only see this after compiling, and nothing warns you.

I ran into it adding subnormal samples to frexp's OpInfo in #194110 -- the new samples turned two CUDA inductor tests red.

Inductor lowers frexp to `triton_helpers.frexp`, which does all the work in floating point:

https://github.com/pytorch/pytorch/blob/9286a168d99c114b4370cc755e6cd075a76822ef/torch/_inductor/runtime/triton_helpers.py#L793-L803

```python
@triton.jit
def frexp(x):
    # TODO(isuruf): use inline_asm_elementwise here
    zero = x == 0
    not_finite = libdevice.isinf(x).to(tl.int1) | libdevice.isnan(x).to(tl.int1)
    special = zero | not_finite
    safe_x = tl.where(special, 1.0, x)
    y = libdevice.ilogb(safe_x) + 1
    exponent = tl.where(special, 0, y)
    mantissa = tl.where(zero, 0, tl.where(not_finite, x, libdevice.ldexp(safe_x, -y)))
    return mantissa, exponent
```

CUDA flushes float32 subnormals to zero, so the value gets eaten somewhere in there. I didn't chase down which line -- both `x == 0` and the `ldexp` would give you the zero mantissa we're seeing.

float64 is fine, which fits: `test_comprehensive_frexp_cuda_float64` passes while the float32 one fails. ROCm and XPU pass the same inputs too.

## Minimal reproducer

```python
import torch

@torch.compile
def f(x):
    return torch.frexp(x)

# float32 subnormals, 1.4e-45 to 1.8e-43
x = torch.tensor([1, 2, 3, 7, 127], dtype=torch.int32, device="cuda").view(torch.float32)

print("eager:   ", torch.frexp(x))
print("compiled:", f(x))
```

Both lines should print what eager and CPU give:

```
mantissa  [0.5, 0.5, 0.75, 0.875, 0.9921875]
exponent  [-148, -147, -147, -146, -142]
```

The compiled one comes back with the mantissa all zeros.

Fair warning: I don't have a CUDA machine, so I wrote that script from the CI failure below rather than running it.

### Error logs

No exception, it's a silent wrong answer. What I actually have is two CI failures on #194110:

- `test/inductor/test_torchinductor_opinfo.py::TestInductorOpInfoCUDA::test_comprehensive_frexp_cuda_float32`
- `test/inductor/test_op_dtype_prop.py::TestCaseCUDA::test_op_dtype_propagation_frexp_cuda_float32`

```
Mismatched elements: 5 / 5 (100.0%)
Greatest absolute difference: 0.9921875 at index (4,) (up to 1.5e-05 allowed)
Greatest relative difference: 1.0 at index (0,) (up to 1.3e-05 allowed)
The failure occurred for item [0]
```

Item `[0]` is the mantissa -- index `(4,)` wanted `0.9921875` and got `0`, index `(0,)` wanted `0.5` and got `0`.

If it helps: pulling the exponent and mantissa out of the bit pattern with integer ops sidesteps this entirely, since integers don't get flushed. That's what #194110 does for MPS, where Apple GPUs flush float32 subnormals the same way. Happy to do the same here if that's the direction you'd want.

### Versions

From CI on `main` at 9286a168d99c114b4370cc755e6cd075a76822ef. I work on macOS/MPS, so `collect_env.py` off my machine wouldn't tell you anything useful about the configuration that's broken -- the failing jobs are `linux-jammy-cuda13.0-py3.10-gcc11` and `linux-jammy-cuda13.2-py3.10-gcc11`.

---

Written with assistance from an AI coding assistant. I've reviewed the analysis and the CI failures it's based on; as noted above, I haven't been able to run the reproducer myself since I don't have a CUDA device.

cc @ptrblck @msaroufim @eqy @tinglvv @nWEIdia @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #195041

Fixes #195007

Inductor lowers torch.frexp to triton_helpers.frexp, which computed the
result in floating point via libdevice.ilogb/ldexp. Triton compiles the
CUDA libdevice bitcode with __CUDA_FTZ=1, so __nv_ilogbf flushes float32
subnormal inputs to zero and returns FP_ILOGB0 (INT_MIN), poisoning both
the mantissa and the exponent. float64 was unaffected because FTZ only
applies to float32. The old code also dropped the sign of -0.0 in the
mantissa via `tl.where(zero, 0, ...)`.

The helper now decomposes the IEEE-754 bit pattern with pure integer ops:
bitcast to a signed integer, extract the exponent/fraction fields, force
the exponent field to BIAS-1 to get a mantissa in [0.5, 1), and rebase
the integer exponent. Subnormals are normalized by converting the
fraction integer to float, which is exact (the fraction fits in the
significand) and always produces a normal number, so FTZ cannot touch
it. Zero, inf, and NaN pass through verbatim with exponent 0, matching
eager and preserving the sign of -0.0. As a side effect, fp16/bf16 now
work natively under triton.codegen_upcast_to_fp32=False, where the old
code failed to compile (libdevice.ilogb has no fp16 overload).

An alternative was inline_asm_elementwise (per the old TODO in the
helper), but that is CUDA-specific; this helper is shared by the CUDA,
ROCm, and XPU Triton backends, and the integer decomposition is portable
to all of them with no libdevice dependency at all.

Test Plan:

New regression test covering fp32/fp64 subnormal boundaries, +-0.0, and
the smallest normals, comparing mantissa bit patterns via .view(int) to
catch sign-of-zero loss; verified it fails against the old
implementation. Also verified bit-exactness against eager on exhaustive
sweeps out-of-band: all fp32 subnormals, all 65536 fp16 and bf16 bit
patterns (both upcast configs), and tens of millions of random fp32/fp64
bit patterns.

```
python test/inductor/test_cuda_repro.py -k frexp
python test/inductor/test_torchinductor_opinfo.py -k frexp
python test/inductor/test_op_dtype_prop.py -k frexp
python test/inductor/test_cpu_repro.py -k frexp
lintrunner -a torch/_inductor/runtime/triton_helpers.py test/inductor/test_cuda_repro.py
```

All pass (6+2+2+1 tests OK, lint clean) on CUDA (B200).

Authored with Claude Code.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by anijain2305** [comment]
@claude review this PR

**Review by anijain2305 — APPROVED** [review]
